modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/jdbc/JdbcThinTransactionalSelfTest.java

```diff
@@ -0,0 +1,204 @@
+import org.apache.ignite.testframework.junits.common.GridCommonAbstractTest;
+import org.apache.ignite.transactions.TransactionConcurrency;
+import org.junit.Test;
+
+import static java.sql.Connection.TRANSACTION_NONE;
+import static java.sql.Connection.TRANSACTION_READ_COMMITTED;
+import static java.sql.Connection.TRANSACTION_READ_UNCOMMITTED;
+import static java.sql.Connection.TRANSACTION_REPEATABLE_READ;
+import static java.sql.Connection.TRANSACTION_SERIALIZABLE;
+import static java.sql.ResultSet.CLOSE_CURSORS_AT_COMMIT;
+import static java.sql.ResultSet.CONCUR_READ_ONLY;
+import static java.sql.ResultSet.HOLD_CURSORS_OVER_COMMIT;
+import static java.sql.ResultSet.TYPE_FORWARD_ONLY;
+import static org.apache.ignite.testframework.GridTestUtils.assertThrows;
+
+/** */
+public class JdbcThinTransactionalSelfTest extends GridCommonAbstractTest {
+    /** URL. */
+    private static final String URL = "jdbc:ignite:thin://127.0.0.1";
+
+    /** {@inheritDoc} */
+    @Override protected IgniteConfiguration getConfiguration(String igniteInstanceName) throws Exception {
+        IgniteConfiguration cfg = super.getConfiguration(igniteInstanceName);
+
+        cfg.getTransactionConfiguration().setTxAwareQueriesEnabled(true);
+        cfg.getSqlConfiguration().setQueryEnginesConfiguration(new CalciteQueryEngineConfiguration());
+
+        return cfg;
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void beforeTestsStarted() throws Exception {
+        super.beforeTestsStarted();
+
+        startGrid();
+    }
+
+    /** */
+    @Test
+    public void testInvalidHoldability() throws Exception {
+        try (Connection conn = DriverManager.getConnection(URL)) {
+            List<TestRunnable> checks = Arrays.asList(
+                () -> conn.setHoldability(HOLD_CURSORS_OVER_COMMIT),
+                () -> conn.createStatement(TYPE_FORWARD_ONLY, CONCUR_READ_ONLY, HOLD_CURSORS_OVER_COMMIT),
+                () -> conn.prepareStatement("SELECT * FROM T", TYPE_FORWARD_ONLY, CONCUR_READ_ONLY, HOLD_CURSORS_OVER_COMMIT)
+            );
+
+            assertEquals(CLOSE_CURSORS_AT_COMMIT, conn.getHoldability());
+
+            for (TestRunnable check : checks) {
+                assertThrows(
+                    null,
+                    () -> {
+                        check.run();
+                        return null;
+                    },
+                    SQLException.class,
+                    "Invalid holdability (can't hold cursor over commit)."
+                );
+            }
+        }
+    }
+
+    /** */
+    @Test
+    public void testTransactionConcurrencyProperty() throws Exception {
+        for (TransactionConcurrency txConcurrency : TransactionConcurrency.values()) {
+            String url = URL + "?transactionConcurrency=" + txConcurrency;
+
+            try (Connection conn = DriverManager.getConnection(url)) {
+                assertEquals(txConcurrency, GridTestUtils.getFieldValue(conn, "txConcurrency"));
```

Comment:
Can we use the same approach as testTransactionLabel?
In this case:
1. No reflection required.
2. We also test that txConcurrency is correctly passed to the server.

Reply:
Good catch. Fixed.